src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -45,40 +53,65 @@ public VisualStudioProjectFactory(
         _dynamicFileInfoProviders = fileInfoProviders.AsImmutableOrEmpty();
         _vsixAnalyzerProviderFactory = vsixAnalyzerProviderFactory;
         _analyzerAssemblyRedirectors = analyzerAssemblyRedirectors.AsImmutableOrEmpty();
         _solution2 = solution2;
-    }
 
-    public Task<ProjectSystemProject> CreateAndAddToWorkspaceAsync(string projectSystemName, string language, CancellationToken cancellationToken)
-        => CreateAndAddToWorkspaceAsync(projectSystemName, language, new VisualStudioProjectCreationInfo(), cancellationToken);
+        _uiContextUpdateWorkQueue = new AsyncBatchingWorkQueue<string>(
+            DelayTimeSpan.NearImmediate,
+            UpdateUIContextsAsync,
+            StringComparer.OrdinalIgnoreCase,
+            listenerProvider.GetListener(FeatureAttribute.Workspace),
+            _threadingContext.DisposalToken);
+
+        _initializationTask = _threadingContext.JoinableTaskFactory.RunAsync(
+            async () =>
+            {
+                var cancellationToken = _threadingContext.DisposalToken;
+                // HACK: Fetch this service to ensure it's still created on the UI thread; once this is
+                // moved off we'll need to fix up it's constructor to be free-threaded.
+
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
+                _visualStudioWorkspaceImpl.Services.GetRequiredService<VisualStudioMetadataReferenceManager>();
+
+                _visualStudioWorkspaceImpl.SubscribeExternalErrorDiagnosticUpdateSourceToSolutionBuildEvents();
+                _visualStudioWorkspaceImpl.SubscribeToSourceGeneratorImpactingEvents();
+
+                // Since we're on the UI thread here anyways, use that as an opportunity to grab the
+                // IVsSolution object and solution file path.
+                var solution = await _solution2.GetValueOrNullAsync(cancellationToken).ConfigureAwait(true);
+                _solutionFilePath = solution != null && ErrorHandler.Succeeded(solution.GetSolutionInfo(out _, out var filePath, out _))
+                    ? filePath
+                    : null;
+
+                _vsixAnalyzerProvider = await _vsixAnalyzerProviderFactory.GetOrCreateProviderAsync(cancellationToken).ConfigureAwait(true);
+            });
+    }
 
-    public async Task<ProjectSystemProject> CreateAndAddToWorkspaceAsync(
-        string projectSystemName, string language, VisualStudioProjectCreationInfo creationInfo, CancellationToken cancellationToken)
+    private async ValueTask UpdateUIContextsAsync(ImmutableSegmentedList<string> list, CancellationToken cancellationToken)
     {
-        // HACK: Fetch this service to ensure it's still created on the UI thread; once this is
-        // moved off we'll need to fix up it's constructor to be free-threaded.
+        // This is not cancellable as we have already mutated the solution.
+        cancellationToken = CancellationToken.None;
```

Comment:
Could the batching queue shouldn't be creating cancellable tokens here in the first place, right? Otherwise I'd be worried the queue might internally cancel work and that'd create problems.

Reply:
the batching queue always supports cancellation.  But it will only actually cancel in one of two cases:

1. The CT that it was passed in its constructor gets canceled.  That'st he threading context disposal token.  So it would be ok in that case.
2. someone does `.AddWork(x, cancelExistingWork: true)`.  This doesn't happen in this code afaict.  So you'd never be triggered by that.

Based on this, i think it's fine to actually check the CT and not do the CT.None bit.  You should only get canceled here during shutdown.  In which case you *do* want to stop working.